externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/session/SparkSQLSessionManager.scala

```diff
@@ -50,6 +53,38 @@ class SparkSQLSessionManager private (name: String, spark: SparkSession)
   val operationManager = new SparkSQLOperationManager()
 
   private lazy val singleSparkSession = conf.get(ENGINE_SINGLE_SPARK_SESSION)
+  private lazy val shareLevel = ShareLevel.withName(conf.get(ENGINE_SHARE_LEVEL))
+
+  private lazy val userIsolatedSparkSession = conf.get(ENGINE_USER_ISOLATED_SPARK_SESSION)
+  private lazy val userIsolatedSparkSessionCache = new ConcurrentHashMap[String, SparkSession]()
+
+  private def getOrNewSparkSession(user: String): SparkSession = {
+    if (singleSparkSession) {
+      spark
+    } else {
+      shareLevel match {
+        // it's unnecessary to create a new spark session in connection share level
+        // since the session is only one
+        case CONNECTION => spark
+        case USER => newSparkSession(spark)
+        case GROUP | SERVER if userIsolatedSparkSession => newSparkSession(spark)
+        case GROUP | SERVER =>
+          if (userIsolatedSparkSessionCache.containsKey(user)) {
+            userIsolatedSparkSessionCache.get(user)
+          } else {
+            val newSession = newSparkSession(spark)
+            userIsolatedSparkSessionCache.put(user, newSession)
```

Comment:
There may be a problem of concurrent session creation here, but it should be no problem to change to guava cache.

Reply:
it's ok for different users in parallel

Reply:
If synchronization is not used here, is it possible for different sessions of the same user to create `SparkSession` at the same time?

Reply:
I see your concern but it is not good practice.. The GROUP/SERVER is for different and same/different session. The case of same user with different session should be use USER share level.